src/token/api/src/types/tx_record.rs

```diff
@@ -1,86 +1,51 @@
-use crate::types::{Operation, TransactionStatus, TxId};
 use candid::{CandidType, Deserialize, Principal};
 use ic_canister::ic_kit::ic;
 use ic_helpers::tokens::Tokens128;
 
+use crate::account::Account;
+use crate::types::{Operation, Timestamp, TransactionStatus, TxId};
+
 #[derive(Deserialize, CandidType, Debug, Clone)]
 pub struct TxRecord {
-    pub caller: Option<Principal>,
+    pub caller: Principal,
     pub index: TxId,
-    pub from: Principal,
-    pub to: Principal,
+    pub from: Account,
+    pub to: Account,
     pub amount: Tokens128,
     pub fee: Tokens128,
-    pub timestamp: u64,
+    pub timestamp: Timestamp,
     pub status: TransactionStatus,
     pub operation: Operation,
+    pub memo: Option<u64>,
 }
 
 impl TxRecord {
     pub fn transfer(
         index: TxId,
-        from: Principal,
-        to: Principal,
+        from: Account,
+        to: Account,
         amount: Tokens128,
         fee: Tokens128,
+        memo: Option<u64>,
+        created_at_time: Option<Timestamp>,
```

Comment:
why add a created at time, without a check? Better to use ic::time?

Reply:
The `created_at_time` is taken from the `TransferArgs` and is optional, I just used the `timestamp` field  to pass in the `created_at_time`

Reply:
we need to check this is within a few minutes of the ic-time, otherwise the transaction should be rejected .. maybe there are hints in the draft implementation.

Reply:
I checked the icrc1 implementation , there are no checks there really.